### PR TITLE
Cleanup quantile.h

### PR DESCRIPTION
**Pull request by RAMitchell**
I found it difficult to understand what is going on in quantile.h, so I went through and removed dead code, tried to simplify some relationships between classes and tried to make as many methods/variables private as possible. This PR should cause no change in behaviour.

In particular:
- WQSummary is removed as everything seems to use WXQSummary
- QuantileSketchTemplate has been collapsed into WXQuantileSketch as it was the only instance being used

Ideally I wanted something like an interface where I could hide away the internals but still use it in a sketching algorithm. This can't happen yet because this part https://github.com/dmlc/xgboost/blob/44469a0ca9375f3f44bc9b827a38e0eeed37d475/src/tree/updater_basemaker-inl.h#L366 
accesses internal members in many places.

### CODE REVIEW

**Comment by trivialfis** [comment]
Plan to add unit tests?

File: src/common/quantile.h [hunk 1]
```diff
@@ -9,21 +9,23 @@
...
 #include <cstring>
-#include <algorithm>
 #include <iostream>
+#include <vector>
 
 namespace xgboost {
 namespace common {
+
 /*!
  * \brief experimental wsummary
```
Comment (trivialfis) [inline src/common/quantile.h]: We don't call it experimental anymore.

File: src/common/hist_util.cc [hunk 2]
```diff
@@ -234,17 +233,13 @@ void DenseCuts::Build(DMatrix* p_fmat, uint32_t max_num_bins) {
...
 
-  unsigned const nstep =
-      static_cast<unsigned>((info.num_col_ + nthread - 1) / nthread);
-  unsigned const ncol = static_cast<unsigned>(info.num_col_);
-  sketchs.resize(info.num_col_);
-  for (auto& s : sketchs) {
-    s.Init(info.num_row_, 1.0 / (max_num_bins * kFactor));
-  }
+  unsigned const nstep = static_cast<unsigned>((info.num_col_ + nthread - 1) /
+  nthread); unsigned const ncol = static_cast<unsigned>(info.num_col_);
```
Comment (trivialfis) [inline src/common/hist_util.cc]: That's a weird way to format code.

File: src/common/quantile.h [hunk 3]
```diff
@@ -418,18 +204,21 @@ struct WXQSummary : public WQSummary<DType, RType> {
           size_t i = bid;
           RType maxdx2 = src.data[end].RMaxPrev() * 2;
           for (; k < n; ++k) {
-            RType dx2 =  2 * ((k * mrange) / n + begin);
+            RType dx2 = 2 * ((k * mrange) / n + begin);
             if (dx2 >= maxdx2) break;
             while (i < end &&
-                   dx2 >= src.data[i + 1].rmax + src.data[i + 1].rmin) ++i;
+                   dx2 >= src.data[i + 1].rmax + src.data[i + 1].rmin)
+              ++i;
```
Comment (trivialfis) [inline src/common/quantile.h]: Add a `{}`.

File: src/common/quantile.h [hunk 4]
```diff
@@ -686,19 +447,18 @@ class QuantileSketchTemplate {
    * \param maxn maximum number of data points can be feed into sketch
    * \param eps accuracy level of summary
    */
-  inline void Init(size_t maxn, double eps) {
```
Comment (trivialfis) [inline src/common/quantile.h]: Great that we can remove the `Init` method.

File: tests/cpp/data/test_quantile.cc [hunk 5]
```diff
@@ -0,0 +1,40 @@
+// Copyright by Contributors
+
+#include <gtest/gtest.h>
+#include "../../../src/common/quantile.h"
+
+namespace xgboost {
+namespace common {
+TEST(Quantile, Basic) {
```
Comment (trivialfis) [inline tests/cpp/data/test_quantile.cc]: Do we need a null test?

File: src/common/quantile.h [hunk 6]
```diff
@@ -444,150 +233,124 @@ struct WXQSummary : public WQSummary<DType, RType> {
...
+        ++b;
       } while (b != b_end);
     }
-    CHECK(dst == data + size) << "bug in combine";
+    this->size = dst - data;
+    const RType tol = 10;
+    RType err_mingap, err_maxgap, err_wgap;
+    this->FixError(&err_mingap, &err_maxgap, &err_wgap);
+    if (err_mingap > tol || err_maxgap > tol || err_wgap > tol) {
+      LOG(INFO) << "mingap=" << err_mingap << ", maxgap=" << err_maxgap
```
Comment (trivialfis) [inline src/common/quantile.h]: Is it necessary to log this?

File: src/common/quantile.h [hunk 7]
```diff
@@ -444,150 +233,124 @@ struct WXQSummary : public WQSummary<DType, RType> {
...
+        *dst = Entry(b->rmin + aprev_rmin, b->rmax + a->RMaxPrev(), b->wmin,
+                     b->value);
+        bprev_rmin = b->RMinNext();
+        ++dst;
+        ++b;
       }
     }
     if (a != a_end) {
-      RType bprev_rmax = (b_end - 1)->rmax;
+      RType brmax = (b_end - 1)->rmax;
```
Comment (trivialfis) [inline src/common/quantile.h]: Please change it to a more verbose name.

File: src/common/quantile.h [hunk 8]
```diff
@@ -408,7 +193,8 @@ struct WXQSummary : public WQSummary<DType, RType> {
     }
```
Comment (trivialfis) [inline src/common/quantile.h]: > CHECK(nbig < n) << "quantile: too many large chunk";

I recently run into this error when weight is sampled from a Gaussian distribution.  Not sure what's happening here.

File: src/common/hist_util.cc (hunk 2 above)
Comment (RAMitchell) [reply src/common/hist_util.cc]: I just auto formatted with clang-format. Will clean it up.

File: src/common/quantile.h (hunk 7 above)
Comment (RAMitchell) [reply src/common/quantile.h]: I actually have no idea what any of this code does, it's just copy pasted. I will spend a bit more time learning the algorithms to see if I can clean it up more.

**Comment by trivialfis** [comment]
Recently I have been learning boost accumulator library.   We might want to borrow some concept from there.

**Pull request closed without merging**